src/awkward/_reducers.py

```diff
@@ -314,16 +313,16 @@ def apply(cls, array, parents, outlength):
             raise ak._errors.wrap_error(
                 ValueError(f"cannot compute the product (ak.prod) of {array.dtype!r}")
             )
-        result = array.nplike.empty(
-            cls.maybe_double_length(array.dtype.type, outlength),
-            dtype=cls.return_dtype(array.dtype),
-        )
         if array.dtype == np.bool_:
+            result = array.nplike.empty(
+                cls.maybe_double_length(array.dtype.type, outlength),
+                dtype=np.dtype(np.bool_),
+            )
             assert parents.nplike is array.nplike
             array._handle_error(
                 array.nplike[
                     "awkward_reduce_prod_bool",
-                    array.dtype.type,
+                    result.dtype.type,
```

Comment:
My only suggestion, feel free to revert.

Reply:
Yes, absolutely. I don't see how it worked without this.

Reply:
With anything but `np.bool_` as the first argument, `nplike.__getitem__` should have been unable to find the ctypes function.

```python
>>> import awkward as ak, numpy as np
>>> numpy = ak.nplikes.nplike_of(None)
>>> numpy["awkward_reduce_prod_bool", np.bool_, np.int64, np.int64]
<NumpyKernel awkward_reduce_prod_bool, bool, int64, int64>
>>> numpy["awkward_reduce_prod_bool", np.int64, np.int64, np.int64]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/irishep/awkward/awkward/nplikes.py", line 495, in __getitem__
    return NumpyKernel(ak._cpu_kernels.kernel[name_and_types], name_and_types)
KeyError: ('awkward_reduce_prod_bool', <class 'numpy.int64'>, <class 'numpy.int64'>, <class 'numpy.int64'>)
```

Oh well! It's correct now and works now.

I'll merge it.